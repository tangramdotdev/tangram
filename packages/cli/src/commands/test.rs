use crate::Cli;
use tangram_client as tg;
use tangram_error::Result;

/// Build the target named "test" from the specified package.
#[derive(Debug, clap::Args)]
#[command(verbatim_doc_comment)]
pub struct Args {
	/// If this flag is set, then the command will exit immediately instead of waiting for the build's output.
	#[arg(short, long)]
	pub detach: bool,

	/// The path to the executable in the artifact to run.
	#[arg(long)]
	pub executable_path: Option<tg::Path>,

	/// If this flag is set, the package's lockfile will not be updated.
	#[arg(long)]
	pub locked: bool,

	/// The package to build.
	#[arg(short, long, default_value = ".")]
	pub package: tg::Dependency,

	/// The retry strategy to use.
	#[arg(long, default_value_t)]
	pub retry: tg::build::Retry,
}

impl Cli {
	pub async fn command_test(&self, args: Args) -> Result<()> {
		// Create the build args.
		let args = super::build::Args {
			detach: args.detach,
			locked: args.locked,
			no_tui: false,
			output: None,
			package: args.package,
			retry: args.retry,
			target: "test".to_owned(),
		};

		// Build!
		self.command_build(args).await?;

		Ok(())
	}
}
