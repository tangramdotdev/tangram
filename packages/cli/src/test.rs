use crate::Config;
use futures::{Future, FutureExt as _};
use std::{panic::AssertUnwindSafe, sync::Arc};
use tangram_client as tg;
use tangram_temp::Temp;
use url::Url;

pub async fn test<F, Fut>(tg: &'static str, f: F)
where
	F: FnOnce(Arc<tokio::sync::Mutex<Context>>) -> Fut,
	Fut: Future<Output = ()> + Send,
{
	// Create the context.
	let context = Arc::new(tokio::sync::Mutex::new(Context::new(tg)));

	// Run the test and catch a panic if one occurs.
	let result = AssertUnwindSafe(f(context.clone())).catch_unwind().await;

	// Handle the result.
	let context = context.lock().await;
	match result {
		Ok(()) => {
			// If there was no panic, then gracefully shutdown the servers.
			for server in &context.servers {
				let mut process = server.process.lock().await;
				unsafe { libc::kill(process.id().unwrap().try_into().unwrap(), libc::SIGINT) };
				let result = process.wait().await;
				if let Ok(status) = result {
					if !status.success() {}
				}
			}
		},
		Err(error) => {
			// If there was a panic, then forcefully shutdown the servers and resume unwinding.
			for server in &context.servers {
				let mut process = server.process.lock().await;
				unsafe { libc::kill(process.id().unwrap().try_into().unwrap(), libc::SIGKILL) };
				process.wait().await.ok();
			}
			std::panic::resume_unwind(error);
		},
	}
}

pub struct Context {
	servers: Vec<Arc<Server>>,
	tg: &'static str,
}

pub struct Server {
	config: Config,
	process: tokio::sync::Mutex<tokio::process::Child>,
	temp: Arc<Temp>,
	tg: &'static str,
}

impl Context {
	#[must_use]
	pub fn new(tg: &'static str) -> Self {
		Self {
			servers: Vec::new(),
			tg,
		}
	}

	pub async fn spawn_server(&mut self) -> tg::Result<Arc<Server>> {
		let server = Server::new(self.tg).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}

	pub async fn spawn_server_with_config(&mut self, config: Config) -> tg::Result<Arc<Server>> {
		let server = Server::with_config(self.tg, config).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}

	pub async fn spawn_server_with_temp_and_config(
		&mut self,
		config: Config,
	) -> tg::Result<Arc<Server>> {
		let server = Server::with_config(self.tg, config).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}
}

impl Server {
	async fn new(tg: &'static str) -> tg::Result<Self> {
		let config = Config::default();
		Self::with_config(tg, config).await
	}

	async fn with_config(tg: &'static str, config: Config) -> tg::Result<Self> {
		// Create a temp and create the config and data paths.
		let temp = Arc::new(Temp::new());
		tokio::fs::create_dir_all(temp.path())
			.await
			.map_err(|source| tg::error!(!source, "failed to create the directory"))?;
		let config_path = temp.path().join(".config/tangram/config.json");
		tokio::fs::create_dir_all(config_path.parent().unwrap())
			.await
			.map_err(|source| tg::error!(!source, "failed to create the config directory"))?;
		let config_json = serde_json::to_vec_pretty(&config).unwrap();
		tokio::fs::write(&config_path, config_json)
			.await
			.map_err(|source| tg::error!(!source, "failed to write the config"))?;
		let data_path = temp.path().join(".tangram");
		tokio::fs::create_dir_all(&data_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to create the data directory"))?;

		// Create the command.
		let log_path = temp.path().join(".tangram/log");
		let stdout = tokio::fs::OpenOptions::new()
			.create(true)
			.write(true)
			.truncate(true)
			.open(&log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?
			.into_std()
			.await;
		let stderr = tokio::fs::OpenOptions::new()
			.create(true)
			.write(true)
			.truncate(true)
			.open(&log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?
			.into_std()
			.await;
		let mut command = tokio::process::Command::new(tg);
		command.stdout(stdout);
		command.stderr(stderr);
		command.arg("--config");
		command.arg(&config_path);
		command.arg("--path");
		command.arg(data_path);
		command.arg("serve");

		// Spawn the process.
		let process = command.spawn().unwrap();
		let process = tokio::sync::Mutex::new(process);

		// Create the server.
		let server = Self {
			config,
			process,
			temp,
			tg,
		};

		Ok(server)
	}

	#[must_use]
	pub fn tg(&self) -> tokio::process::Command {
		let mut command = tokio::process::Command::new(self.tg);
		let config_path = self.temp.path().join(".config/tangram/config.json");
		let data_path = self.temp.path().join(".tangram");
		command
			.arg("--config")
			.arg(config_path)
			.arg("--path")
			.arg(data_path)
			.arg("--mode")
			.arg("client");
		command
	}

	#[must_use]
	pub fn config(&self) -> &Config {
		&self.config
	}

	#[must_use]
	pub fn process(&self) -> &tokio::sync::Mutex<tokio::process::Child> {
		&self.process
	}

	#[must_use]
	pub fn url(&self) -> Url {
		let path = self.temp.path().join(".tangram/socket");
		let path = path.to_str().unwrap();
		let path = urlencoding::encode(path);
		format!("http+unix://{path}").parse().unwrap()
	}

	pub async fn print_log(&self) -> tg::Result<()> {
		let log_path = self.temp.join(".tangram/log");
		let mut log_file = tokio::fs::File::open(log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?;
		let mut stderr = tokio::io::stderr();
		tokio::io::copy(&mut log_file, &mut stderr).await.ok();
		Ok(())
	}
}

#[macro_export]
macro_rules! assert_output_success {
	($output:expr) => {
		let output = &$output;
		if !output.status.success() {
			let mut stderr = tokio::io::stderr();
			stderr.write_all(&output.stderr).await.unwrap();
			stderr.flush().await.unwrap();
		}
		assert!(output.status.success());
	};
}
pub use assert_output_success;
