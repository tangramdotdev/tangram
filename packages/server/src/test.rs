use crate::{Config, Server};
use futures::{Future, FutureExt as _};
use std::{panic::AssertUnwindSafe, sync::Arc};
use tangram_temp::Temp;

pub async fn test<F, Fut>(f: F)
where
	F: FnOnce(Arc<tokio::sync::Mutex<Context>>) -> Fut,
	Fut: Future<Output = ()> + Send,
{
	// Create the context.
	let context = Arc::new(tokio::sync::Mutex::new(Context::new()));

	// Run the test and catch a panic if one occurs.
	let result = AssertUnwindSafe(f(context.clone())).catch_unwind().await;

	// Handle the result.
	let context = context.lock().await;
	for server in &context.servers {
		server.stop();
		server.wait().await;
	}
	match result {
		Ok(()) => (),
		Err(error) => {
			std::panic::resume_unwind(error);
		},
	}
}

#[derive(Default)]
pub struct Context {
	temps: Vec<Temp>,
	servers: Vec<Server>,
}

impl Context {
	#[must_use]
	pub fn new() -> Self {
		Self::default()
	}

	pub async fn start_server(&mut self) -> Server {
		let temp = Temp::new();
		let config = Config::with_path(temp.path().to_owned());
		self.temps.push(temp);
		let server = Server::start(config).await.unwrap();
		self.servers.push(server.clone());
		server
	}

	pub async fn start_server_with_temp_and_config(
		&mut self,
		temp: Temp,
		config: Config,
	) -> Server {
		self.temps.push(temp);
		let server = Server::start(config).await.unwrap();
		self.servers.push(server.clone());
		server
	}
}
