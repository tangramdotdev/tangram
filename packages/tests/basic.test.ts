import { describe, expect, test } from "bun:test";
import { $ } from "bun";
import { extractBuildId, getTestPackage, startServer } from "./setup.ts";

describe("server", () => {
	test("reports health", async () => {
		await using server = await startServer({ dbConnections: 2 });

		const server_health = await $`tg --config ${server.handle.configPath} server health`.text();
		expect(server_health).toMatchSnapshot();
	});
	
	test("builds a basic package", async () => {
		await using server = await startServer();

		const five = getTestPackage("five");

		const buildOutput = await $`tg --config ${server.handle.configPath} build ${five} --quiet`.text().then((t) => t.trim());

		expect(buildOutput).toBe("5");
	});
	
	test("builds a package with a path dependency", async () => {
		await using server = await startServer();

		const importsFiveByPath = getTestPackage("import_five_by_path");

		 const buildOutput = await $`tg --config ${server.handle.configPath} build ${importsFiveByPath} --quiet`.text().then((t) => t.trim());

		 expect(buildOutput).toBe("6");
	});
});

describe("remote", () => {
	test("cache hit after push", async () => {
		await using remote = await startServer({ registry: true });

		await using originalServer = await startServer({ remotePath: remote.handle.serverPath });

		const five = getTestPackage("five");

		// Build the package.
		const originalBuildOutput = await $`tg --config ${originalServer.handle.configPath} build ${five}`.text();
		const originalBuildId = extractBuildId(originalBuildOutput);
		expect(originalBuildId).toBeDefined();

		// Push the build.
		await $`tg --config ${originalServer.handle.configPath} push ${originalBuildId}`.quiet();

		// Stop the server.
		await originalServer.handle.stop();

		// Start a new server.
		await using freshServer = await startServer({ remotePath: remote.handle.serverPath });

		// Build the package on the new server.
		await $`tg --config ${freshServer.handle.configPath} build ${five}`.text();

		// Get the original build ID.
		const getOrigBuildId = await $`tg --config ${freshServer.handle.configPath} get ${originalBuildId}`.text();

		// The returned output should include the build ID.
		expect(getOrigBuildId).toMatch(originalBuildId as string);
	});

	test("push object to remote", async () => {
		await using remote = await startServer({ registry: true });

		await using server = await startServer({ remotePath: remote.handle.serverPath });

		const five = getTestPackage("five");

		// Check in the test package.
		const fivePackageId = await $`tg --config ${server.handle.configPath} checkin ${five}`.text().then((t) => t.trim());

		// Store the output of `tg get` on that ID.
		const localGetFiveOutput = await $`tg --config ${server.handle.configPath} get ${fivePackageId}`.text().then((t) => t.trim());

		// Push the test package
		await $`tg --config ${server.handle.configPath} push ${fivePackageId}`.quiet();

		// Stop the local server.
		await server.handle.stop();
		
		// Store the output of `tg get` on the remote using that ID.
		const remoteGetFiveOutput = await $`tg --config ${remote.handle.configPath} get ${fivePackageId}`.text().then((t) => t.trim());

		// They should match.
		expect(localGetFiveOutput).toBe(remoteGetFiveOutput);
	});

	test("builds from pushed tag name", async () => {
		await using remote = await startServer({ registry: true });

		await using originalServer = await startServer({ remotePath: remote.handle.serverPath });

		const five = getTestPackage("five");

		// Tag the package.
		await $`tg --config ${originalServer.handle.configPath} tag five ${five}`.quiet();
		
		// Push the tag.
		await $`tg --config ${originalServer.handle.configPath} push five`.quiet();

		// Stop the server.
		await originalServer.handle.stop();

		// Start a fresh server.
		await using freshServer = await startServer({ remotePath: remote.handle.serverPath });

		// Build using just the tag.
		const buildOutput = await $`tg --config ${freshServer.handle.configPath} build five --quiet`.text().then((t) => t.trim());
		
		// We should have received the correct output without ever passing the path to this server.
		expect(buildOutput).toBe("5");
	});
	
	test("build alternate target after pushing build", async () => {
		await using remote = await startServer({ registry: true });

		await using originalServer = await startServer({ remotePath: remote.handle.serverPath });

		const twoTargets = getTestPackage("two_targets");

		// Tag the package.
		await $`tg --config ${originalServer.handle.configPath} tag twoTargets ${twoTargets}`.quiet();
		
		// Push the tag.
		await $`tg --config ${originalServer.handle.configPath} push twoTargets`.quiet();

		// Build the package.
		const originalBuildOutput = await $`tg --config ${originalServer.handle.configPath} build twoTargets#five`.text();
		const originalBuildId = extractBuildId(originalBuildOutput);
		expect(originalBuildId).toBeDefined();

		// Push the build.
		await $`tg --config ${originalServer.handle.configPath} push ${originalBuildId}`.quiet();

		// Stop the server.
		await originalServer.handle.stop();

		// Start a new server.
		await using freshServer = await startServer({ remotePath: remote.handle.serverPath });

		// Build the other target package on the new server. The tag should retrieve the package, but start a new build.
		let otherTargetOutput = await $`tg --config ${freshServer.handle.configPath} build twoTargets#six --quiet`.text().then((t) => t.trim());

		// The build should succeed.
		expect(otherTargetOutput).toBe("6");
	});
})
