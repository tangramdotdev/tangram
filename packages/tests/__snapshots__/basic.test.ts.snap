// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`server reports health 1`] = `
"{
  "builds": {
    "created": 0,
    "dequeued": 0,
    "started": 0
  },
  "database": {
    "available_connections": 2
  },
  "file_descriptor_semaphore": {
    "available_permits": 1024
  },
  "version": "0.0.0"
}
"
`;
